app/controllers/network_services_controller.go

```diff
@@ -375,6 +379,159 @@ func ensureMasqueradeIptablesRule(masqueradeAll bool, podCidr string) error {
+	jumpArgs := []string{"-m", "ipvs", "-j", hairpinChain}
+	err = iptablesCmdHandler.AppendUnique("nat", "POSTROUTING", jumpArgs...)
+
+	// Key is a string that will match iptables.List() rules
+	// Value is a string[] with arguments that iptables transaction functions expect
+	rulesNeeded := make(map[string][]string, 0)
+
+	// Generate the rules that we need
+	for svcName, svcInfo := range nsc.serviceMap {
+		if nsc.globalHairpin || svcInfo.hairpin {
+			for _, ep := range nsc.endpointsMap[svcName] {
+				rule, ruleArgs := hairpinRuleFrom(svcInfo.clusterIP.String(), ep.ip)
+				rulesNeeded[rule] = ruleArgs
+			}
+		}
+	}
+
+	// Apply the rules we need
+	for _, ruleArgs := range rulesNeeded {
+		err = iptablesCmdHandler.AppendUnique("nat", hairpinChain, ruleArgs...)
+		if err != nil {
+			return errors.New("Failed to apply hairpin iptables rule: " + err.Error())
+		}
+	}
+
+	rulesFromNode, err := iptablesCmdHandler.List("nat", hairpinChain)
+	if err != nil {
+		return errors.New("Failed to get rules from iptables chain \"" +
+			hairpinChain + "\": " + err.Error())
+	}
+
+	// Delete invalid/outdated rules
+	for _, ruleFromNode := range rulesFromNode {
+		_, ruleIsNeeded := rulesNeeded[ruleFromNode]
+		if !ruleIsNeeded {
+			args := strings.Fields(ruleFromNode)
+			if len(args) > 2 {
+				args = args[2:] // Strip "-A CHAIN_NAME"
+
+				err = iptablesCmdHandler.Delete("nat", hairpinChain, args...)
+				if err != nil {
+					glog.Errorf("Unable to delete hairpin rule \"%s\" from chain %s: %e", ruleFromNode, hairpinChain, err)
+				} else {
+					glog.Info("Deleted invalid/outdated hairpin rule \"%s\" from chain %s", ruleFromNode, hairpinChain)
+				}
+			} else {
+				glog.Infof("Not removing invalid hairpin rule \"%s\" from chain %s", ruleFromNode, hairpinChain)
+			}
+		}
+	}
+
+	return nil
+}
+
+func hairpinRuleFrom(serviceIP string, endpointIP string) (string, []string) {
+	// TODO: Factor hairpinChain out
+	hairpinChain := "KUBE-ROUTER-HAIRPIN"
+
+	ruleArgs := []string{"-s", endpointIP + "/32", "-d", endpointIP + "/32",
+		"-m", "ipvs", "--vaddr", serviceIP, "-j", "SNAT", "--to-source", serviceIP}
+
+	// Trying to ensure this matches iptables.List()
+	ruleString := "-A " + hairpinChain + " -s " + endpointIP + "/32" + " -d " +
+		endpointIP + "/32" + " -m ipvs" + " --vaddr " + serviceIP + " -j SNAT" +
+		" --to-source " + serviceIP
+
+	return ruleString, ruleArgs
+}
+
+func deleteHairpinIptablesRules() error {
+	// TODO: Factor these variables out
+	hairpinChain := "KUBE-ROUTER-HAIRPIN"
+	hasHairpinChain := false
+
+	iptablesCmdHandler, err := iptables.New()
+	if err != nil {
+		return errors.New("Failed to initialize iptables executor" + err.Error())
+	}
+
+	// TODO: Factor this static jump rule out
+	jumpArgs := []string{"-m", "ipvs", "-j", hairpinChain}
+	err = iptablesCmdHandler.Delete("nat", "POSTROUTING", jumpArgs...)
+	if err != nil {
+		glog.Errorf("Unable to delete hairpin jump rule from chain \"POSTROUTING\": %e", err)
```

Comment:
We can return from here. If there is reference from nat POSTROUTING chain to hairpin chain, delete hairpin chain will fail anyways.